File: programmers/뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기_2.py
def solution(numbers):
    answer = [-1] * len(numbers)
    stack = []

    # 순서대로 탐색
    for i in range(len(numbers)):
        while stack and numbers[stack[-1]] < numbers[i]:
            answer[stack.pop()] = numbers[i]
        stack.append(i)

    while stack:
        answer[stack.pop()] = -1

    return answer


print(solution([2, 3, 3, 5]))
print(solution([9, 1, 5, 3, 6, 2]))

# 참고
# https://velog.io/@wlals425315/%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%A8%B8%EC%8A%A4-154539-%EB%92%A4%EC%97%90-%EC%9E%88%EB%8A%94-%ED%81%B0-%EC%88%98-%EC%B0%BE%EA%B8%B0

File: programmers/뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기_2.java
import java.util.Arrays;
import java.util.Stack;

public class Main {
    public static void main(String[] args){
        Solution solution = new Solution();
        int[] maps = {2, 3, 3, 5};
        System.out.println(Arrays.toString(solution.solution(maps)));
    }
}

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = Arrays
                .stream(new int[numbers.length])
                .map(v -> v - 1)
                .toArray();

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < numbers.length; i++) {
            while (!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
                answer[stack.pop()] = numbers[i];
            }
            stack.add(i);
        }

        while (!stack.isEmpty()) {
            answer[stack.pop()] = -1;
        }

        return answer;
    }
}
